cmd/gateway/s3x/gateway-s3x-ledger-store.go

```diff
@@ -0,0 +1,152 @@
+	l   *Ledger          //a cache of the values in datastore and ipfs
+
+	locker     bucketLocker //a locker to protect buckets from concurrent access (per bucket)
+	plocker    bucketLocker //a locker to protect MultipartUploads from concurrent access (per upload ID)
+	mapLocker  sync.Mutex   //a lock to protect the l.Buckets map from concurrent access
+	pmapLocker sync.Mutex   //a lock to protect the l.MultipartUploads map from concurrent access
+}
+
+func newLedgerStore(ds datastore.Batching, dag pb.NodeAPIClient) (*ledgerStore, error) {
+	ls := &ledgerStore{
+		ds:  namespace.Wrap(ds, dsPrefix),
+		dag: dag,
+		l: &Ledger{
+			Buckets:          make(map[string]*LedgerBucketEntry),
+			MultipartUploads: make(map[string]*MultipartUpload),
+		},
+	}
+	return ls, nil
+}
+
+func (ls *ledgerStore) object(ctx context.Context, bucket, object string) (*Object, error) {
+	b, err := ls.getBucketLoaded(ctx, bucket)
+	if err != nil {
+		return nil, err
+	}
+	objs := b.GetBucket().Objects
+	if objs == nil {
+		return nil, ErrLedgerObjectDoesNotExist
+	}
+	h, ok := b.GetBucket().Objects[object]
+	if !ok {
+		return nil, ErrLedgerObjectDoesNotExist
+	}
+	return ipfsObject(ctx, ls.dag, h)
+}
+
+//ObjectInfo returns the ObjectInfo of the object.
+func (ls *ledgerStore) ObjectInfo(ctx context.Context, bucket, object string) (*ObjectInfo, error) {
+	obj, err := ls.object(ctx, bucket, object)
+	if err != nil {
+		return nil, err
+	}
+	return &obj.ObjectInfo, nil
+}
+
+func (ls *ledgerStore) ObjectData(ctx context.Context, bucket, object string) ([]byte, error) {
+	obj, err := ls.object(ctx, bucket, object)
+	if err != nil {
+		return nil, err
+	}
+	return ipfsBytes(ctx, ls.dag, obj.GetDataHash())
+}
+
+// RemoveObject is used to remove a ledger object entry from a ledger bucket entry
+func (ls *ledgerStore) removeObject(ctx context.Context, bucket, object string) error {
+	missing, err := ls.removeObjects(ctx, bucket, object)
+	if err != nil {
+		return err
+	}
+	if len(missing) != 0 {
+		return ErrLedgerObjectDoesNotExist
+	}
+	return nil
+	//todo: gc on ipfs
+}
+
+// removeObjects efficiently remove many objects, returns a list of objects that did not exist.
+func (ls *ledgerStore) removeObjects(ctx context.Context, bucket string, objects ...string) ([]string, error) {
+	b, err := ls.getBucketLoaded(ctx, bucket)
+	if err != nil {
+		return nil, err
+	}
+	if b.Bucket.Objects == nil {
+		return objects, nil
+	}
+
+	missing := []string{}
+	for _, o := range objects {
+		_, ok := b.Bucket.Objects[o]
+		if !ok {
+			missing = append(missing, o)
+			continue
+		}
+		delete(b.Bucket.Objects, o)
+	}
+	_, err = ls.saveBucket(ctx, bucket, b.Bucket)
+	return missing, err
+	//todo: gc on ipfs
+}
+
+//PutObject saves an object by hash into the given bucket
+func (ls *ledgerStore) PutObject(ctx context.Context, bucket, object string, obj *Object) error {
+	defer ls.locker.write(bucket)()
+	return ls.putObject(ctx, bucket, object, obj)
```

Comment:
We might want to do:
```
err := ls.putObject(ctx, bucket, object, obj)
ls.locker.wirte(bucket)()
return err
```

Would this `defer` technically run before `ls.putObject` is finished?

Reply:
Your change is a slight optimization, but otherwise the code is the same, unless putObject panics.

Reply:
good point